src/dbnode/encoding/iterators.go

```diff
@@ -38,21 +39,71 @@ var (
 // iterators is a collection of iterators, and allows for reading in order values
 // from the underlying iterators that are separately in order themselves.
 type iterators struct {
-	values      []Iterator
-	earliest    Iterator
-	earliestIdx int
-	earliestAt  time.Time
-	filtering   bool
-	filterStart time.Time
-	filterEnd   time.Time
+	values             []Iterator
+	earliest           []Iterator
+	earliestAt         time.Time
+	filterStart        time.Time
+	filterEnd          time.Time
+	filtering          bool
+	equalTimesStrategy IterateEqualTimestampStrategy
+
+	// Used for caching reuse of value frequency lookup
+	valueFrequencies map[float64]int
 }
 
 func (i *iterators) len() int {
 	return len(i.values)
 }
 
 func (i *iterators) current() (ts.Datapoint, xtime.Unit, ts.Annotation) {
-	return i.earliest.Current()
+	numIters := len(i.earliest)
+
+	switch i.equalTimesStrategy {
+	case IterateHighestValue:
+		sort.Slice(i.earliest, func(a, b int) bool {
+			currA, _, _ := i.earliest[a].Current()
+			currB, _, _ := i.earliest[b].Current()
+			return currA.Value < currB.Value
+		})
+
+	case IterateLowestValue:
+		sort.Slice(i.earliest, func(a, b int) bool {
+			currA, _, _ := i.earliest[a].Current()
+			currB, _, _ := i.earliest[b].Current()
+			return currA.Value > currB.Value
+		})
+
+	case IterateHighestFrequencyValue:
```

Comment:
lol this feels so magic-y. Do you have a use in mind for it?

Reply:
As per docs its basically if you are using `consistency: all` and don't mind the read unavailability when not all replicas are available to be read from (say perhaps exact match mode for testing correctness for shadow traffic, etc).